### PR TITLE
Add unit tests for `util/Git`

### PR DESCRIPTION
**Pull request by jacobroufa**

### CODE REVIEW

File: tests/unit/src/util/Git.ts [hunk 1]
```diff
@@ -0,0 +1,350 @@
+import * as registerSuite from 'intern!object';
+import * as assert from 'intern/chai!assert';
+import loadModule, { cleanupModuleMocks } from '../../../_support/loadModule';
+import { stub, SinonStub } from 'sinon';
+import Git from '../../../../src/util/Git';
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: use the AMD mapping for this and environment

File: tests/unit/src/util/Git.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+		git.isInitialized = () => true;
+		git.getConfig = stub().withArgs('remote.origin.url')
+			.returns('url');
+
+		git.assert('other_url').then(() => assert.fail(), (error: Error) => {
+			assert.strictEqual(error.message,
+				'Repository mismatch. Expected "url" to be "other_url".');
+		});
+		assert.doesNotThrow(async () => await git.assert('url'));
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: I don't think doesNotThrow wrappers like this work with async yet. Once async/await isn't transpiled it should be ok.

can we switch it to `return git.assert('url');`

File: tests/unit/src/util/Git.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+					'A clone directory must be set');
+			});
+		},
+
+		// logger should be called once internally and once within execSSHAgent
+		async 'Not initialized; logger called 1x, git.url === url'() {
+			const url = 'url';
+			git.isInitialized = () => false;
+			await git.clone(url);
+			assert.isTrue(loggerStub.calledTwice);
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: Let's not test the logger. It's a brittle test as a change to logging in any method will break this test.

Instead, let's test that assert is not called

```
const assertSpy = spy(git, 'assert');
assert.false(assertSpy.called);
```

We should also test that execSSHAgent is called once

File: tests/unit/src/util/Git.ts [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+		}).returns(Promise.resolve('pass'));
+
+		const actual = git.checkout('1.2.3');
+
+		assert.instanceOf(actual, Promise);
+		assert.strictEqual(await actual, 'pass');
+	},
+
+	clone: {
+		'If clone directory is not set; throws'() {
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: can we change `throws` to `rejects` or `eventually rejects`

File: tests/unit/src/util/Git.ts [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+			assert.strictEqual(git.url, url);
+		},
+
+		async 'Properly initialized; logger called 3x, git.url === url'() {
+			const url = 'url';
+			git.isInitialized = () => true;
+			git.assert = stub().withArgs(url)
+				.returns(true);
+			await git.clone(url);
+			assert.isTrue(loggerStub.calledThrice);
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: switch it to spy and test assert and execSSHAgent

File: tests/unit/src/util/Git.ts [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+			assert.strictEqual(git.url, url);
+		}
+	},
+
+	async commit() {
+		const esa: SinonStub = git.execSSHAgent = stub().returns(Promise.resolve());
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: switch to use `const execSSHAgentStub = stub(git, 'execSSHAgent').returns(Promise.resolve())`

File: tests/unit/src/util/Git.ts [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+			delete git.cloneDirectory;
+			git.createOrphan('branch').then(() => assert.fail(), (error: Error) => {
+				assert.equal(error.message,
+					'A clone directory must be set');
+			});
+		},
+
+		async 'promiseExec called twice, logger.info called once'() {
+			await git.createOrphan('branch');
+			assert.isTrue(promiseExecStub.calledTwice);
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: Let's assert that promiseExecStub is passed what we expect

File: tests/unit/src/util/Git.ts [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+			assert.isTrue(setConfig.calledWith('user.email', 'email'));
+		}
+	},
+
+	execSSHAgent: {
+		async '!hasDeployCredentials; logger, promiseSpawn called'() {
+			git.hasDeployCredentials = () => false;
+			await git.execSSHAgent('git', [ 'status' ], { silent: false });
+			assert.isTrue(loggerStub.calledOnce);
+			assert.isTrue(promiseSpawnStub.calledOnce);
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: Assert arguments

File: tests/unit/src/util/Git.ts [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+		},
+
+		async 'hasDeployCredentials; chmodSync, promiseExec called'() {
+			git.hasDeployCredentials = () => true;
+			await git.execSSHAgent('git', [ 'status' ], { silent: false });
+			assert.isTrue(chmodSyncStub.calledOnce);
+			assert.isTrue(promiseExecStub.calledOnce);
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: assert arguments

File: tests/unit/src/util/Git.ts [hunk 10]
```diff
@@ -105,7 +97,7 @@ registerSuite({
 			assert.strictEqual(error.message,
 				'Repository mismatch. Expected "url" to be "other_url".');
 		});
-		assert.doesNotThrow(async () => await git.assert('url'));
+		assert.doesNotThrow(() => git.assert('url'));
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: This doesn't work. This needs to be:

```typescript
return git.assert('url')
```

File: tests/unit/src/util/Git.ts [hunk 11]
```diff
@@ -121,38 +113,42 @@ registerSuite({
 	},
 
 	clone: {
-		'If clone directory is not set; throws'() {
+		'If clone directory is not set; eventually rejects'() {
 			delete git.cloneDirectory;
 			git.clone('url').then(() => assert.fail(), (error: Error) => {
```
Comment (devpaul) [inline tests/unit/src/util/Git.ts]: This async function creates a promise that needs to be returned from the test

File: tests/unit/src/util/Git.ts (hunk 10 above)
Comment (jacobroufa) [inline tests/unit/src/util/Git.ts]: Sorry, I misunderstood your previous request, and took the `return` to be *from* the anonymous function, not the test itself. Now I understand that if it returns an assertion and that doesn't pass, the test will fail. 🤦‍♂️

**Pull request merged**